design/BufferOperations.md

```diff
@@ -0,0 +1,148 @@
+The results are initialized in the "pending" state and transition at Javascript task boundary to the "available" state when the implementation can determine the GPU is done using the buffer.
+Calling `mapRead` or `mapWrite` puts the buffer in the mapped state.
+No operations are allowed in a buffer in that state except additional calls to `mapRead` or `mapWrite` and calls to `unmap`.
+In particular a mapped buffer cannot be used in a `WebGPUCommandBuffer` given to `WebGPUQueue.submit`.
+The following must be true or a validation error occurs for `mapWrite` (resp. `mapRead`):
+
+ - The buffer must have been created with the `WebGPUBufferUsage.MAP_WRITE` (resp. `WebGPUBufferUsage.MAP_READ`) usage.
+ - `offset + size` must not overflow and be at most the size of the buffer
+ - Depending on the design of memory barriers, the buffer must be, or allowed to be in the `WebGPUBufferUsage.MAP_WRITE` (resp. `WebGPUBufferUsage.MAP_READ`) usage.
+
+Then a mapped buffer can be unmapped with:
+
+```
+partial interface WebGPUBuffer {
+    void unmap();
+};
+```
+
+This operation invalidates all the `WebGPUMappedMemory` created from the buffer and puts the buffer in the unmapped state.
+The buffer must be in the mapped state otherwise a validation error occurs when `unmap` is called.
+
+### WebGPUMappedMemory
+
+`WebGPUMappedMemory` is an object representing a mapped region of a buffer that's both pollable and promise-like.
+
+It can be in one of three states: pending, available and invalidated.
+
+The pollable interface is:
+
+```
+partial interface WebGPUMappedMemory {
+    bool isPending();
+    ArrayBuffer getPointer();
+};
+```
+
+`isPending` return true if the object is in the pending state, false otherwise.
+`getPointer` returns an ArrayBuffer representing the buffer data if the object is in the available state, null otherwise.
+
+`WebGPUMappedMemory` is also `then`-able, meaning that it acts like a Javascript `Promise`:
+
+```
+partial interface WebGPUMappedMemory {
+    Promise then(WebGPUMappedMemorySuccessCallback success,
+                 optional WebGPUMappedMemoryErrorCallback error);
+};
+```
+
+This acts like a `Promise<ArrayBuffer>.then` that is resolved on the Javascript task boundary in which the implementation detects the GPU is done with the buffer.
+On that boundary:
+
+ - The `WebGPUMappedMemory` goes in the available state.
+ - If the `WebGPUMappedMemory` was created via `WebGPUBuffer.mapWrite`, its content is cleared to 0.
+ - `success` is called with the content of the memory as an argument.
+
+If `success` hasn't been called when the WebGPUMappedMemory gets invalidated (meaning the object is still in the pending state), `error` is called instead. When `WebGPUMappedMemory` goes from the available state to the invalidated state, the `ArrayBuffer` for its content gets neutered. The return value of `then` acts like the return value of `Promise.then`.
+
```

Comment:
Thanks, I now remember we mentioned something like that in a meeting. A potential solution was for `.then` to act like a `Promise<void>` and just signal that the buffer is now in the available state? What do you think of something like this?

Reply:
It can also be `Promise<WebGPUMappedMemory>`. Then `getPointer` can still return null (or neutered AB, or whatever).

Reply:
Although if `getPointer` is going to return a neutered AB, there's no reason the Promise can't resolve to a neutered AB as well.

Reply:
My questions was more about what "validity" means in the context of buffers and the event loop.  For readback, if multiple pieces of code `.then` the promise, will the ArrayBuffer only be non-neutered when the first `.then` is invoked or when all of them are invoked? When will `WebGPUMappedMemory` go from the available state to the invalidated state?

Reply:
It never happens automatically, only on `unmap`.

Reply:
Thank you for the clarification, @kainino0x .  That means if the developer calls `.then` on the WebGPUMappedMemory object multiple times and unmaps it on the first callback, all of the subsequent callbacks would get a neutered `ArrayBuffer`.  I presume this is necessary because we want to unmap the buffer as soon as we can to keep the pipeline going.  I guess this is a case of "To bad, so sad" for the subsequent callbacks.  :-)

Reply:
Perhaps we could sidestep this problem by using callbacks directly. There is no timing problem because the earliest the callback could be run is at the next micro task boundary. It also solves the problem of code reusability (due to Promises not being able to be re-used once they are fulfilled).

Reply:
I think there would be pushback from the TAG, but it's a possibility.
(I don't know that they would like our "thenable" thing any better.)